### PR TITLE
Add D'Esopo-Papes's algorithm for shortest paths

### PR DESCRIPTION
**Pull request by eulerkochy**
This algorithm outperforms Bellman-ford and Dijkstra in majority cases and is very easy to implement. Sadly, it doesn't work for negative cycles.

### CODE REVIEW

**Comment by sbromberger** [comment]
This is really interesting - I hadn't heard of this algorithm before.

Before we can merge it, we need tests. And preferably some benchmarks on some SNAP graphs.

**Comment by jpfairbanks** [comment]
I found this source for a discussion of pro/con with D'Esopo-Pape's algorithm. http://web.mit.edu/dimitrib/www/SLF.pdf

I'm worried about the "sometimes exponential time" part. Is there any way to determine if that behavior occurs?

**Comment by simonschoelly** [comment]
Here is a short (maybe not the best) benchmark
```
julia> using BenchmarkTools, LightGraphs, SNAPDatasets

julia> fb = SimpleGraph{Int32}(loadsnap(:facebook_combined));

julia> @btime dijkstra_shortest_paths(fb, Int32(1));
  1.685 ms (47 allocations: 337.63 KiB)

julia> @btime bellman_ford_shortest_paths(fb, Int32(1));
  608.008 μs (103 allocations: 146.48 KiB)

julia> @btime DEsopoPape_shortest_path(fb, Int32(1));
  458.649 μs (18 allocations: 111.73 KiB)
```

File: src/shortestpaths/desopo-pape.jl [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    struct DEposoPapeState{T, U}
+
+An [`AbstractPathState`](@ref) designed for D`Esopo-Pape shortest-path calculations.
+"""
+struct DEsopoPapeState{T <:Real, U <: Integer} <: AbstractPathState
+    parents::Vector{U}
+    dists::Vector{T}
+end
+
+function DEsopoPape_shortest_path(g::AbstractGraph,
```
Comment (simonschoelly) [inline src/shortestpaths/desopo-pape.jl]: Functions in Julia usually start with a lower case letter.
Also, considering that `dijkstra_shortests_paths` and `bellman_ford_shortest_paths` both end with an `s`, this function probably should also do that.

File: src/shortestpaths/desopo-pape.jl [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+An [`AbstractPathState`](@ref) designed for D`Esopo-Pape shortest-path calculations.
+"""
+struct DEsopoPapeState{T <:Real, U <: Integer} <: AbstractPathState
+    parents::Vector{U}
+    dists::Vector{T}
+end
+
+function DEsopoPape_shortest_path(g::AbstractGraph, 
+    src::U,
```
Comment (simonschoelly) [inline src/shortestpaths/desopo-pape.jl]: Don't take the eltype from the source vertex, but take it from the graph `g` and then convert `src` to that eltype at the beginning of the functions, so that we can avoid time consuming convertions inside the loops.

File: src/shortestpaths/desopo-pape.jl [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    distmx::AbstractMatrix{T} = weights(g)) where T <: Real where U <: Integer
+    
+    nvg = nv(g)
+    dists = fill(typemax(T), nvg)
+    parents = zeros(U, nvg)
+    state = fill(2, nvg)
```
Comment (simonschoelly) [inline src/shortestpaths/desopo-pape.jl]: consider using a `Vector{UInt8}` or `Vector{Int8}` instead of the Vector{Int} here.

File: src/shortestpaths/desopo-pape.jl [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    q = Array{U,1}()
```
Comment (simonschoelly) [inline src/shortestpaths/desopo-pape.jl]: `q = Vector{U}()`

File: src/shortestpaths/desopo-pape.jl [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    push!(q, src)
+    dists[src] = 0
+    
+    while !isempty(q)
```
Comment (simonschoelly) [inline src/shortestpaths/desopo-pape.jl]: You have a lot of array access here, so you can use `@inbounds` here.

File: src/shortestpaths/desopo-pape.jl (hunk 1 above)
Comment (eulerkochy) [reply src/shortestpaths/desopo-pape.jl]: Ok, cool

File: src/shortestpaths/desopo-pape.jl (hunk 2 above)
Comment (eulerkochy) [reply src/shortestpaths/desopo-pape.jl]: Ok!

File: src/shortestpaths/desopo-pape.jl (hunk 3 above)
Comment (eulerkochy) [reply src/shortestpaths/desopo-pape.jl]: for `state` , right?

File: src/shortestpaths/desopo-pape.jl (hunk 4 above)
Comment (eulerkochy) [reply src/shortestpaths/desopo-pape.jl]: ok!

File: src/shortestpaths/desopo-pape.jl (hunk 3 above)
Comment (simonschoelly) [reply src/shortestpaths/desopo-pape.jl]: Yes

File: src/shortestpaths/desopo-pape.jl (hunk 5 above)
Comment (eulerkochy) [reply src/shortestpaths/desopo-pape.jl]: I was skeptical initially about this. Is this necessary?

File: src/shortestpaths/desopo-pape.jl (hunk 2 above)
Comment (eulerkochy) [reply src/shortestpaths/desopo-pape.jl]: >take it from the graph `g` 

As in?

File: src/shortestpaths/desopo-pape.jl (hunk 5 above)
Comment (simonschoelly) [reply src/shortestpaths/desopo-pape.jl]: No, but it can make stuff much faster, because otherwise you would have a branch for every array access. Might be a good idea to only add `@inbounds`, when you are finished with writing an algorithm, so until then, you see when you make an error.

File: src/shortestpaths/desopo-pape.jl (hunk 2 above)
Comment (eulerkochy) [reply src/shortestpaths/desopo-pape.jl]: `edgetype` of `g`?

Comment (simonschoelly) [reply src/shortestpaths/desopo-pape.jl]: You either write `g::AbstractGraph{U}` in the signature or you use `U = eltype(g)` later in the code. In Both cases you should write `src::Integer`.

Comment (simonschoelly) [reply src/shortestpaths/desopo-pape.jl]: No, `eltype` which for graphs is actually the type of the vertex. For other collections, eltype is usual the type of the stored element.

Comment (eulerkochy) [reply src/shortestpaths/desopo-pape.jl]: cool, i used `eltype(g)`

**Comment by eulerkochy** [comment]
The current implementation with `@inbounds` is somewhat faster
image

**Comment by simonschoelly** [comment]
> The current implementation with `@inbounds` is somewhat faster
> image

Uhm, but you are not comparing the new results on your computer with my old ones, right?

File: src/shortestpaths/desopo-pape.jl [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    U = eltype(g)
+    nvg = nv(g)
+    dists = fill(typemax(T), nvg)
+    parents = zeros(U, nvg)
+    state = Vector{UInt8}()
+    for i=1:nvg
+        push!(state, 2)
+    end
+    q = Vector{U}()
+    push!(q, src)
```
Comment (simonschoelly) [inline src/shortestpaths/desopo-pape.jl]: You could replace
```
q = Vector{U}()
push!(q, src)
```
by `q = U[src]`

File: src/shortestpaths/desopo-pape.jl [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    src::Integer,
+    distmx::AbstractMatrix{T} = weights(g)) where T <: Real
+    U = eltype(g)
+    nvg = nv(g)
+    dists = fill(typemax(T), nvg)
+    parents = zeros(U, nvg)
+    state = Vector{UInt8}()
+    for i=1:nvg
+        push!(state, 2)
+    end
```
Comment (simonschoelly) [inline src/shortestpaths/desopo-pape.jl]: Repeately pushing is not that performant. You could instead just write `state = fill(UInt8(2), nvg)`

File: src/shortestpaths/desopo-pape.jl [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+    q = Vector{U}()
+    push!(q, src)
+    @inbounds dists[src] = 0
```
Comment (simonschoelly) [inline src/shortestpaths/desopo-pape.jl]: We will get into trouble here, if `src` is not in the range `1:nvg`. You could for example add a test for that at the beginning of the function, and if it not the case, then throw a `DomainError`.

File: test/shortestpaths/desopo-pape.jl [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    add_edge!(G, 3, 5)
+    add_edge!(G, 3, 4)
+    add_edge!(G, 4, 5)
+    
+    for g in testgraphs(G)
+        y = @inferred(desopo_pape_shortest_paths(g, 1, m))
+        @test z.parents == [0, 1, 1, 3, 3]
+        @test z.dists == [0, 2, 2, 3, 4]
+    end
+end
```
Comment (simonschoelly) [inline test/shortestpaths/desopo-pape.jl]: The good think in this case is, that we already have algorithms such as `dijikstra_shortests_paths`, that should return the same distances. (I think `parents` could be different though). So you could create a random graph and random weights and compare if the distances of both algorithms are approximately the same. (You can use the `≈` operator for that).

Also, you should test for some special cases, such as empty graphs, graphs with just one vertex, disconnected graphs and graphs with self-loops.

It is not very popular at the moment in the LightGraphs code base, but you can use `@testset` to subdivide your testset even more and give some meaningful names, so that we now, what each part is testing.

File: src/shortestpaths/desopo-pape.jl [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+    
+    while !isempty(q)
+        @inbounds u = q[1]
+        popfirst!(q)
```
Comment (simonschoelly) [inline src/shortestpaths/desopo-pape.jl]: I don't know exactly how performat `popfirst!` and `pushfirst!` for vectors are. It might be worth to use a queue from the `DataStructures.jl` package. If you do that, you should benchmark, if that makes your code faster or slower.

Comment (simonschoelly) [reply src/shortestpaths/desopo-pape.jl]: Also `popfirst!` does also return the element that pops, so you can shorten these two lines to one.

File: test/shortestpaths/desopo-pape.jl (hunk 9 above)
Comment (eulerkochy) [reply test/shortestpaths/desopo-pape.jl]: Excellent idea!

File: src/shortestpaths/desopo-pape.jl (hunk 8 above)
Comment (eulerkochy) [reply src/shortestpaths/desopo-pape.jl]: Sure!

**Comment by eulerkochy** [comment]
> Uhm, but you are not comparing the new results on your computer with my old ones, right?

Nah, I had them separately tested out on my machine.

File: src/shortestpaths/desopo-pape.jl (hunk 10 above)
Comment (eulerkochy) [reply src/shortestpaths/desopo-pape.jl]: Both `popfirst!` and `pushfirst!` are `O(1)` operations, so it shouldn't matter if I use a `queue` or not. I'll try benchmarking and update you about the result. 

> Also `popfirst!` does also return the element that pops, so you can shorten these two lines to one.

Ok, nice observation!

File: src/shortestpaths/desopo-pape.jl (hunk 7 above)
Comment (eulerkochy) [reply src/shortestpaths/desopo-pape.jl]: Ok!

File: src/shortestpaths/desopo-pape.jl (hunk 10 above)
Comment (eulerkochy) [reply src/shortestpaths/desopo-pape.jl]: Having `Deque` in the code is slowing it down a lot! I suggest we keep the implementation as it is. Moreover, it has been mentioned in the documentation of `Deque`, 
>Benchmark shows that the performance of Deque is comparable to Vector on push!, and is noticeably faster on unshift! (by about 30% to 40%).

File: src/shortestpaths/desopo-pape.jl [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    parents::Vector{U}
+    dists::Vector{T}
+end
+
+function desopo_pape_shortest_paths(g::AbstractGraph, 
+    src::Integer,
+    distmx::AbstractMatrix{T} = weights(g)) where T <: Real
+    U = eltype(g)
+    nvg = nv(g)
+    (src in 1:nvg) || throw(DomainError("src should be in between 1 and $nvg"))
```
Comment (simonschoelly) [inline src/shortestpaths/desopo-pape.jl]: The constructor for `DomainError` is either `DomainError(val)` or `DomainError(val, msg)` so in your case, you probably should write `DomainError(src, "....")`.

File: src/shortestpaths/desopo-pape.jl [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+    dists = fill(typemax(T), nvg)
+    parents = zeros(U, nvg)
+    state = Vector{Int8}()
+    state = fill(Int8(2), nvg)
+    q = U[src]
+    @inbounds dists[src] = 0
+    
+    @inbounds while !isempty(q)
+        @inbounds u = popfirst!(q)
```
Comment (simonschoelly) [inline src/shortestpaths/desopo-pape.jl]: Putting `@inbounds` in front of the while-block, disables the boundchecks for all array access inside the block (as long as they don't happen inside a function called there, in which case, there are special rules), so you don't have to write `@inbounds` inside the loop.

**Comment by simonschoelly** [comment]
That is an impressive amount of of tests. I think structuring with nested `@testgraph` really helps making this readable, so I think that is something that should be adapted in other places in LightGraphs.

I think, you don't have a test for graphs with self-loops yet. As these graphs often require special consideration in our algorithms, I think we need a test with such graphs.

Can you also add a docstring to your function?

File: src/shortestpaths/desopo-pape.jl (hunk 11 above)
Comment (eulerkochy) [reply src/shortestpaths/desopo-pape.jl]: Ok, onto it!

File: src/shortestpaths/desopo-pape.jl (hunk 12 above)
Comment (eulerkochy) [reply src/shortestpaths/desopo-pape.jl]: Done!

File: src/LightGraphs.jl [hunk 13]
```diff
@@ -20,7 +20,7 @@ import SparseArrays: blockdiag, sparse
 import Base: adjoint, write, ==, <, *, ≈, convert, isless, issubset, union, intersect,
             reverse, reverse!, isassigned, getindex, setindex!, show,
             print, copy, in, sum, size, eltype, length, ndims, transpose,
-            join, iterate, eltype, get, Pair, Tuple, zero
+            join, iterate, eltype, get, Pair, Tuple, zero, push!, popfirst!, pushfirst!
```
Comment (simonschoelly) [inline src/LightGraphs.jl]: It shouldn't be necessary to import  `push! , `popfirst! ` and `pushfirst` - the reason that we import the other functions from `Base` here is that we overload these functions for our types and for that you have to explicitly import them.

Comment (eulerkochy) [reply src/LightGraphs.jl]: Ok, cool! I'm scraping them!

**Comment by simonschoelly** [comment]
Cool, from my side, this can be merged.

Maybe at one point in the future we can figure out if there is a faster double ended queue, but this can also be done for multiple functions in this library.

If you don't understand why I requested some changes, just ask.

If you have any questions

**Comment by eulerkochy** [comment]
It's cool, I changed only after reflection why those necessary at that time. Thanks for your guidance!

**Comment by simonschoelly** [comment]
But now Travis fails with some weird error message, do the test work for you locally?

**Comment by eulerkochy** [comment]
Yeah, they are working! I was also wondering why are the CI test failing?

File: test/shortestpaths/desopo-pape.jl [hunk 14]
```diff
@@ -0,0 +1,168 @@
+@testset "D'Esopo-Pape" begin
+    @test "Generic tests for graphs" begin
```
Comment (simonschoelly) [inline test/shortestpaths/desopo-pape.jl]: Should be `@testset`

Comment (eulerkochy) [reply test/shortestpaths/desopo-pape.jl]: XD! I also found that out!

**Comment by eulerkochy** [comment]
And finally, the tests passed! :grinning:

File: src/shortestpaths/desopo-pape.jl (hunk 8 above)
Comment (sbromberger) [reply src/shortestpaths/desopo-pape.jl]: I don't think we check this anywhere else:
```
julia> g = Graph(10,20)
{10, 20} undirected simple Int64 graph

julia> dijkstra_shortest_paths(g, 100)
ERROR: BoundsError: attempt to access 10-element Array{Int64,1} at index [100]
```

File: src/shortestpaths/desopo-pape.jl [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+        u = popfirst!(q)
+        state[u] = 0
+        
+        for v in outneighbors(g, u)
+            alt = dists[u] + distmx[u, v]
+            if (dists[v] > alt)
+                dists[v] = alt
+                parents[v] = u
+                
+                if (state[v] == 2)
```
Comment (sbromberger) [inline src/shortestpaths/desopo-pape.jl]: no parentheses here, please.

File: src/shortestpaths/desopo-pape.jl [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+                    state[v] = 1
+                    push!(q, v)
+                elseif (state[v] == 0)
```
Comment (sbromberger) [inline src/shortestpaths/desopo-pape.jl]: no parentheses here, please.
